client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -525,6 +525,116 @@ private RetryAndThrowConditionally createRetryConfig(final ControllerImplConfig
+            new ControllerClientTagger(client, timeoutMillis).withTag(requestId, "removeSubscriber", scope, streamName)
+                    .removeSubscriber(ModelHelper.decode(scope, streamName, subscriber), callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(x -> {
+            switch (x.getStatus()) {
+                case FAILURE:
+                    log.warn(requestId, "Failed to update stream: {}", streamName);
+                    throw new ControllerFailureException("Failed to add subscriber: " + subscriber);
+                case SCOPE_NOT_FOUND:
+                    log.warn(requestId, "Scope not found: {}", scope);
+                    throw new IllegalArgumentException("Scope does not exist: " + scope);
+                case STREAM_NOT_FOUND:
+                    log.warn(requestId, "Stream does not exist: {}", streamName);
+                    throw new IllegalArgumentException("Stream does not exist: " + streamName);
+                case SUBSCRIBER_NOT_FOUND:
+                    log.warn(requestId, "Subscriber does not exist: {} for stream {}/{}", subscriber, scope, streamName);
+                    throw new IllegalArgumentException("Subscriber does not exist: " + subscriber);
+                case SUCCESS:
+                    log.info(requestId, "Successfully removed subscriber {} from stream: {}/{}", subscriber, scope, streamName);
+                    return true;
+                case UNRECOGNIZED:
+                default:
+                    throw new ControllerFailureException("Unknown return status adding subscriber " + subscriber
+                            + " " + x.getStatus());
+            }
+        }).whenComplete((x, e) -> {
+            if (e != null) {
+                log.warn(requestId, "removeSubscriber {} for stream {}/{} failed: ", subscriber, scope, streamName, e);
+            }
+            LoggerHelpers.traceLeave(log, "removeSubscriber", traceId, subscriber, requestId);
+        });
+    }
+
+    @Override
+    public CompletableFuture<List<String>> getSubscribersForStream(String scope, String streamName) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkNotNull(scope, "scopeName");
+        Preconditions.checkNotNull(streamName, "stream");
+        final long requestId = requestIdGenerator.get();
+        long traceId = LoggerHelpers.traceEnter(log, "getSubscribersForStream", streamName);
+        final CompletableFuture<StreamSubscribers> result = this.retryConfig.runAsync(() -> {
+            RPCAsyncCallback<StreamSubscribers> callback = new RPCAsyncCallback<>(requestId, "getSubscribersForStream", scope, streamName);
+            client.withDeadlineAfter(timeoutMillis, TimeUnit.MILLISECONDS).getSubscribersForStream(ModelHelper.createStreamInfo(scope, streamName), callback);
+            return callback.getFuture();
+        }, this.executor);
+        return result.thenApply(subscribers -> {
```

Comment:
I primarily added this API for testing. Client does not invoke it ever. Hence did not add a separate status for this.
But it can be added if you thing this could be invoked by the Client.

Reply:
the response structure for this api is a defined as a new protobuf message. so you have the opportunity to include a status and a response list. 
and you have pointed out an important thing -- that the pattern for streamnotfound is not consistently followed across apis stream apis.. please create a new issue in backlog for these (unrelated to current change set) to include a status with each such api where we can set the appropriate status or make use of grpc status runtime to indicate such failures..